### DISCUSSION TITLE
How to get the editor context from IDE

### DISCUSSION THREAD

**Discussion by yuqizhou77** [community/community#146061] [Copilot Conversations]
### Select Topic Area

Question

### Body

I am following https://github.com/copilot-extensions/blackbeard-extension to build my first extension. I am happy to see that if I @mention the participant, it can give me response. And if I add file using #file, the file will be added as copilot_references that I can get from request payload. However, if I just choose the current file sending message like this,
![image](https://github.com/user-attachments/assets/63f2f8b1-7a27-49f0-bb0a-e7326ec9f9f8)
It seems that the file data won't be included in the payload. Similarly, for #terminalSelection, I didn't see any data sent either.

May I know how to enable this? According to https://code.visualstudio.com/docs/copilot/copilot-chat#_extensioncontributed-chat-participants, it sounds like an authorization request will be sent before accessing local context, but I did not see this.

**Comment by AronSoto** [comment]
Hey @yuqizhou77

I see you're diving into building your first extension, that's awesome! 🚀 Let me help you out here.

> [!TIP]
> To include the editor context (like file data or terminal selections) in the request payload, you need to ensure that the **Copilot Chat API** has the proper permissions enabled for accessing local context. Based on the [documentation](https://code.visualstudio.com/docs/copilot/copilot-chat#_extensioncontributed-chat-participants), an authorization request might be required before accessing this local data. Here's a checklist you can follow:

1. **Enable Permissions**: Check if your extension explicitly requests access to editor or terminal context. If not, you might need to modify the settings or your extension's code to include this.

2. **Use `#file` Tag**: You mentioned that adding files via `#file` works—this suggests the extension is correctly configured to process this type of data. For "Current file" references, make sure your code listens for and properly serializes this input in the payload.

3. **Debugging the Payload**: Use a debugger or log the request payload to confirm whether any editor context is being added. If not, double-check the API hooks you're using in the extension.

4. **Consult Examples**: The [Blackbeard Extension](https://github.com/copilot-extensions/blackbeard-extension) is a solid start. Ensure you're calling the necessary methods to fetch and attach editor context properly.

If you're still stuck, feel free to share a snippet of your extension code or the payload structure you're working with. I'd be happy to help further!

Cheers! ✌️

**Comment by iufor** [comment]
Hello
To access editor context in an IDE using natural language processing (NLP), you utilize APIs or SDKs that expose editor data as structured objects. For instance, in IntelliJ IDEA, the Editor class provides insights like caret position or selected text. By extracting these contextual elements, NLP models can analyze user intent, process commands, or suggest actions dynamically, enabling advanced text understanding and intelligent editing features.

**Reply by yuqizhou77** [reply]
I am building a GitHub copilot extension, so if my understanding is correct, the request should be from GitHub Copilot. And my service need to handle the payload, and send response back. So I think the local context editor should be something from Copilot?

**Comment by yuqizhou77** [comment]
Looks like a new permission just being introduced
![image](https://github.com/user-attachments/assets/ce81d189-8976-4798-b398-e5674bb25de0)
